common/persistence/persistence-tests/dynamicRateLimitTest.go

```diff
@@ -0,0 +1,173 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package persistencetests
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	enumsspb "go.temporal.io/server/api/enums/v1"
+	replicationspb "go.temporal.io/server/api/replication/v1"
+	"go.temporal.io/server/common/backoff"
+	"go.temporal.io/server/common/config"
+	"go.temporal.io/server/common/debug"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/persistence/client"
+)
+
+type (
+	DynamicRateLimitSuite struct {
+		TestBase
+		// override suite.Suite.Assertions with require.Assertions; this means that s.NotNil(nil) will stop the test,
+		// not merely log an error
+		*require.Assertions
+
+		ctx    context.Context
+		cancel context.CancelFunc
+
+		rateLimiter *client.HealthRequestRateLimiterImpl
+	}
+)
+
+func (s *DynamicRateLimitSuite) SetupSuite() {
+	healthSignals := persistence.NewHealthSignalAggregatorImpl(
+		30*time.Second,
+		1000,
+		metrics.NoopMetricsHandler,
+		dynamicconfig.GetIntPropertyFn(50),
+		s.Logger,
+	)
+
+	rateLimiterParams := dynamicconfig.GetMapPropertyFn(map[string]interface{}{
+		"enabled":              true,
+		"refreshInterval":      "5ms",
+		"latencyThreshold":     100,
+		"errorThreshold":       0.2,
+		"rateBackoffStepSize":  0.3,
+		"rateIncreaseStepSize": 0.1,
+	})
+
+	rateLimiter := client.NewHealthRequestRateLimiterImpl(
+		healthSignals,
+		func() float64 { return float64(200) },
+		rateLimiterParams,
+		log.NewNoopLogger(),
+	)
+
+	s.TestBase = NewTestBaseWithCassandra(&TestBaseOptions{FaultInjection: &config.FaultInjection{Rate: 0.0}})
+	s.TestBase.PersistenceHealthSignals = healthSignals
+	s.TestBase.PersistenceRateLimiter = rateLimiter
+	s.rateLimiter = rateLimiter
+
+	s.TestBase.Setup(nil)
+}
+
+func (s *DynamicRateLimitSuite) TearDownSuite() {
+	s.TearDownWorkflowStore()
+}
+
+func (s *DynamicRateLimitSuite) SetupTest() {
+	// Have to define our overridden assertions in the test setup. If we did it earlier, s.T() will return nil
+	s.Assertions = require.New(s.T())
+	s.ctx, s.cancel = context.WithTimeout(context.Background(), 30*time.Second*debug.TimeoutMultiplier)
+}
+
+func (s *DynamicRateLimitSuite) TearDownTest() {
+	s.cancel()
+}
+
+func (s *DynamicRateLimitSuite) TestDynamicRateLimiterAdjustsRate() {
```

Comment:
I don't think we need an integration test (with actual persistence setup) for testing the behavior?
And this can be a unit test in like persistence/client/health_request_rate_limiter_test.go?

Reply:
I think using the integration test is a little more complete, since it also tests the signal aggregation part. Is there a compelling reason not to have it as an integration test?

Reply:
It's a bit confusing to me since we are not actually testing the persistence part, so logically it doesn't need to be an integration test.

Reply:
That makes sense. I was mostly just looking to reuse the test setup logic to get a DataStoreFactory with fault injection, rather than setting a bunch of mocked responses.